internals/overlord/tlsstate/manager.go

```diff
@@ -0,0 +1,382 @@
+	"sync"
+	"time"
+)
+
+var (
+	// systemTime can be faked during testing.
+	systemTime = time.Now
+
+	// idCertValidity sets the x509 identity key validity period. The
+	// certificate is managed by the TLS manager as it is not directly
+	// supported by a crypto.Signer interface, as its an implementation
+	// detail related to TLS (not a signing key). Expiry of this
+	// certificate does not change the identity, but changes the client
+	// pinned identity certificate, forcing the client to re-pair with
+	// the server.
+	idCertValidity = 10 * 365 * 24 * time.Hour
+
+	// tlsCertValidity defines how long the in-memory TLS keypairs are
+	// valid before a new keypair is generated. This allows for
+	// TLS session resumption during this period, and keypair re-use
+	// from concurrent requests. The manager restart will always result
+	// in the creation of a new TLS keypair.
+	tlsCertValidity = time.Hour
+
+	// tlsCertRenewWindow defines how long from the actual certificate
+	// expiry we should rather rotate in order to avoid a race between
+	// the expiry and the session handshake completing in time.
+	tlsCertRenewWindow = 60 * time.Second
+
+	// idCertFile is the public x509 certificate, which holds the
+	// identity public key, and self-signed with the identity key.
+	idCertFile = "identity.pem"
+)
+
+type TLSManager struct {
+	// tlsDir is the location of the PEM keypair files.
+	tlsDir string
+	mu     sync.RWMutex
+	// The active in-memory TLS certificate (keypair).
+	tlsCert *tls.Certificate
+	// The identity certificate loaded from disk.
+	idCert *x509.Certificate
+	// The identity key used for signing TLS certificates.
+	signer crypto.Signer
+}
+
+// NewManager create a new TLS keypair manager. The tlsDir must be a
+// directory of which only the leaf element, at most, does not exist. The
+// leaf directory will be created with 0o700 permissions. The signer
+// must implement a crypto.Signer interface. The signer represents the
+// identity key of the machine, container or device. The signer will be
+// used to sign TLS keypairs and its certificate (CA) is included in the
+// TLS certificate chain, allowing a client to pin the identity
+// certificate once trust has been manually established as part of a
+// trust exchange procedure.
+func NewManager(tlsDir string, signer crypto.Signer) (*TLSManager, error) {
+	manager := &TLSManager{
+		tlsDir: tlsDir,
+		signer: signer,
+	}
+	return manager, nil
+}
+
+// GetCertificate returns an identity signed TLS certificate. The certificate chain includes
+// both the TLS leaf certificate, as well as the self signed identity CA certificate. If
+// either the identity or TLS certificate nears expiry, this functions creates new
+// certificates on demand. Note that even if the identity certificate is re-created, this
+// does not mean that the identity key changed (the key itself has no expiry).
+func (m *TLSManager) GetCertificate(*tls.ClientHelloInfo) (*tls.Certificate, error) {
+	var err error
```

Comment:
Why is this needed, instead of the more common `err := foo()` pattern below?

Reply:
This was a leftover line. Removing. Resolving.